### DISCUSSION TITLE
운영체제 - 운영체제 구조와 컴퓨터 구조 - 호현&일구&성은

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#97] [질의 응답]
- 커널에 대해 설명하고, 커널의 역할에 대해 말해주세요.
- 하드웨어에서 인터럽트가 발생하는 경우, 운영체제에서 인터럽트를 처리하는 과정에 대해 설명해주세요.
- 컴퓨터의 기본적인 구성 요소에 대해 설명해주세요.
- 운영체제의 중요한 점은 "다중 프로그래밍 (Multi programming)"이 가능하다는 점입니다. 다중 프로그래밍이란 무엇인가요?
- 명령어 처리 과정에 대해 설명해주세요.

**Comment by seongeun42** [comment]
### ❓️ 커널에 대해 설명하고, 커널의 역할에 대해 말해주세요.

<br>

### 💬 답변
<blockquote>

<img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/014bb48b-7f3a-4f8f-97c9-b65b519b8e10" width="380">

**커널**은 운영체제 중 항상 메모리에 올라가 있는 **운영체제의 핵심 부분**으로써 하드웨어와 응용 프로그램 사이에서 인터페이스를 제공하는 역할을 하며 컴퓨터 자원들을 관리하는 역할을 한다.

- 커널이 하는 일
    - 스케줄링, 메모리 관리, 리소스 할당, 입출력 관리, 파일 시스템 관리, 프로세스 간 통신 등의 일을 함
    - 빠르게 반응해야 하기 때문에 메모리에 항상 실행되는 프로세스로 존재
    - 운영체제의 성능을 좌우함

</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️ 커널의 구성에 대해 말해주세요.

<br>

### 💬 답변
<blockquote>

- 어떻게 기능을 구현하는가에 따라 구분됨
   
 |커널||
|:---:|:---|
|단일형 구조 커널(Monolithic Architecture)|- 초창기 운영체제의 구조 <br> - 커널의 핵심 기능을 구현하는 모듈의 구분 없이 하나로 구성 <br> - 모듈 간 통신 비용이 줄어듦 <br> - 버그나 오류 처리 어려움, 이식성 낮음|
|계층형 구조 커널(Layered Architecture)|- 비슷한 기능을 가진 모듈을 묶어 하나의 계층으로 만듦 <br> - 계층 간 통신을 통해 운영체제 구현 <br> - 버그나 오류 처리, 디버깅하기 더 쉬움|
|마이크로 구조 커널(Micro Architecture)|- 운영체제의 가장 기본적인 기능만 제공 <br> - 모듈은 세분화되어 존재하며, 모듈 간 정보 교환은 프로세스 간 통신을 통해 이루어짐|

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 하드웨어에서 인터럽트가 발생하는 경우, 운영체제에서 인터럽트를 처리하는 과정에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* 인터럽트란 마이크로프로세서(CPU)가 프로그램을 실행하고 있을 때, 입출력 하드웨어 등의 장치에 예외상황이 발생하여 처리가 필요할 경우에 마이크로프로세서에게 알려 처리할 수 있도록 하는 것.
* 인터럽트의 종류
   * 하드웨어 인터럽트
      * Timer, Keyboard 입력, mouse 입력, DMA 등
      * 인터럽트는 시스템 버스를 통해 전달
   * 소프트웨어 인터럽트
      * system call로써 구현
      * 인터럽트 요청 신호에 따라 수행하는 루틴이 인터럽트 처리 프로그램, 즉 ISR 인터럽트 서비스 루틴
      * 인터럽트 벡터(Interrupt vector)는 인터럽트 서비스 루틴의 주소 배열
      * 프로그램이 돌다가 인터럽트 발생 시 발생 전에 실행 중인 그 정보를 저장하고 있어야(스택 영역에) 다음에 다시 돌아와서 이전 프로그램을 실행할 수 있음

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/36f1f4e0-6b0f-40e3-9f3c-f4d0d9f43ec8)

* 인터럽트 처리 과정
   * 명령어 실행단계를 마칠 때마다 cpu는 반복적으로 인터럽트 요청이 있는지 확인
   * 인터럽트가 발생하면, 현재 진행 중인 기계어 코드를 완료
      * 명령어 사이클인 인출과 실행 단계를 일단 마침
   * CPU의 특수레지스터 중, 하이로인터럽트 마스크 비트를 보고 마스크 되면 인터럽트 무시
   * 인터럽트 벡터를 읽고, ISR(인터럽트 서비스 루틴, 인터럽트 핸들러)주소값을 얻음
   * ISR로 점프 한다
      * 이때 PC(Program Counter, IP) 값은 자동 대피 저장
   * 현재 진행중인 프로그램의 레지스터를 대피
   * 해당 코드를 실행
   * 해당 일을 다 처리하면, 대피시킨 레지스터를 복원
   * ISR의 끝에 IRET 명령어에 의해 인터럽트가 해제
   * IRET 명령어가 실행되면, 대피시킨 PC 값을 복원하여 이전 실행 위치로 복원
* 보통 ISR로 들어가면서 CPU코어의 인터럽트 마스크 비트를 자동 설정하고 시작되기 때문에 다른 인터럽트 처리는 대기 상태로 되는 경우가 일반적
   * 따라서 ISR 내에서 다른 인터럽트를 처리하고 싶을 경우 이 마스크 비트를 해제 해야 함

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 컴퓨터의 기본적인 구성 요소에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

|구성||
|:---:|:---|
|CPU|컴퓨터 시스템을 통제하고 프로그램의 연산을 실행 및 처리하는 핵심적인 컴퓨터의 제어 장치|
|Memory|작업에 필요한 프로그램과 수치, 명령, 자료 등의 데이터를 기억하는 기억 장치|
|I/O Device|외부의 데이터를 컴퓨터에 입력하고, 처리한 결과를 사용자가 원하는 형태로 보여주는 입출력 장치|
|메인보드|CPU와 Memory 등 다양한 부품을 가는 선(->버스)으로 연결하는 커다란 판|

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 운영체제의 중요한 점은 "다중 프로그래밍 (Multi programming)"이 가능하다는 점입니다. 다중 프로그래밍이란 무엇인가요?

<br>

### 💬 답변
<blockquote>

* 여러 개의 프로그램을 메모리에 적재하여 실행하는데 하나를 실행하다 대기 상태가 되면 다른 프로그램을 실행한다.
* 하나의 프로그램을 실행하다가 입출력과 관련된 작업이 실행되면 대기 상태가 되고 이 때 다른 프로그램을 실행하는 것이다.
* 일괄 처리 시스템과 달리 입출력 장치가 일을 하는 동안 cpu가 쉬는 것이 아니라 다른 프로그램에 대해서 일을 계속 하기 때문에 cpu 사용 효율이 증가한다.
* 이 시스템에서 운영체제는 작업 스케줄링과 cpu 스케줄링을 진행한다.
   * 작업 스케줄링 : 디스크의 작업 저장소에서 적절한 작업을 메모리로 옮긴다.
   * cpu 스케줄링 : 현재 실행중인 작업이 대기 상태에 들어가면 적절한 작업을 메모리에서 선택해서 cpu에게 전달한다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 일괄 처리 시스템 (batch processing), 시분할 시스템 (time sharing, multitasking), 다중 프로세서 시스템 (multi processor system)

<br>

### 💬 답변
<blockquote>

* 일괄 처리 시스템
   * 문자 그대로 비슷한 작업들을 모아서 일괄 처리 하는 시스템이다.
   * 입출력 장치의 속도가 느려서 cpu가 일을 쉬는 경우가 많았다.
* 시분할 시스템
   * 시분할 시스템은 또 다른 말로 멀티태스킹 이라고도 한다.
   * 시분할 시스템은 이름 그대로 특정 시간을 정해서 그 시간이 지나면 무조건 다음 작업을 실행하는 방식이다.
   * 이 시간이 매우 짧기 때문에 사용자와 컴퓨터는 상호작용을 할 수 있어서 대화식 컴퓨터 시스템 (interactive computer system)이라고 한다.
   * 여러 사용자가 동시에 컴퓨터를 사용할 수 있는데 이 때문에 주기억장치의 용량이 부족할 수 있다.
      * 그래서 가상 메모리를  사용한다.
* 다중 프로세서 시스템
   * 여러 개의 프로세서가 메모리, 클럭 등을 공유하여 사용하는 시스템이다.
   * 여러 개의 프로세서를 다중 프로세서 시스템을 사용하는 것이 여러 개의 단일 프로세서 시스템을 사용하는 것보다 저렴하고 신뢰도가 높다고 한다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 명령어 처리 과정에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* 컴퓨터는 주 기억 장치에 기억되어 있는 명령어를 호출하는 패치과정을 수행하고 해독한다.
* 해독된 결과에 따라서 자료를 이동하고 연산,입출력을 실행하는 방식이다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/611adf9a-7ed9-4339-948f-b3b118714998)

* 호출(Fatch) 사이클
   * 레지스터 PC에서 저장되어 있는 값에 해당하는 주 기억장치로 가서 데이터를 복사해서 레지스터 PC에 저장한다.
   * 명령어가 "JUMP" 경우, PC는 새로운 값
* 해석(Decode) 사이클
   * 레지스터 IR로 가져온 기계어 비트열을 해석한다
   * 연산자의 유형에 따라 피연산자 필드를 적절히 나눈다.
* 실행(Execute) 사이클
   * 레지스터 IR의 비트열에 해당하는 신호를 연산장치에 제공하고,범용 레지스터나 주기억장치에 보내서 기계어 명령어를 수행한다.

</blockquote>

<br>
